primitives/domain/src/lib.rs

```diff
@@ -137,11 +141,14 @@ pub struct MainchainAddress(BoundedVec<u8, ConstU32<MAX_MAINCHAIN_ADDRESS_BYTES>
 
 #[cfg(feature = "serde")]
 impl FromStr for MainchainAddress {
-	type Err = &'static str;
+	type Err = String;
 
 	fn from_str(s: &str) -> Result<Self, Self::Err> {
 		let bytes: Vec<u8> = s.as_bytes().to_vec();
-		let bounded = BoundedVec::try_from(bytes).map_err(|_| "Invalid length")?;
+		let len = bytes.len();
+		let bounded = BoundedVec::try_from(bytes).map_err(|_| {
+			format!("Invalid length: {len}, expected max {MAX_MAINCHAIN_ADDRESS_BYTES}")
+		})?;
```

Comment:
How is this change related to hashing?

Reply:
It's not, I changed the wrong place and left it. Reverted.